### ISSUE TITLE
Rule proposal: warn on unused results in `Promise.all()`

### ISSUE THREAD

**Issue by ulrichstark** [typescript-eslint/typescript-eslint#11717]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Warn when values returned from Promise.all() or Promise.allSettled() are not all destructured, unless the ignored promises resolve to void or undefined. This is useful to avoid potentially costly calls performance wise that are still there after refactoring code.

### Fail Cases

```TypeScript
// Ignored promise returns a meaningful type
const [user] = await Promise.all([fetchUser(), fetchSettings()]);
```

### Pass Cases

```TypeScript
// All results used
const [a, b] = await Promise.all([getA(), getB()]);

// Ignored promise returns void
const [user] = await Promise.all([fetchUser(), logUsageMetrics()]);
```

### Additional Info

This rule flags situations where `Promise.all()` is called with more promises than there are variables to receive their results, meaning some resolved values are ignored. Such code is often unintentional. It may be leftover from a refactor, hide a logic bug, or trigger unnecessary asynchronous work like API calls or database queries whose results are never used.

Because some promises return `void` for side effects, the rule should use type information to avoid false positives. By checking the resolved types of each promise, it can distinguish between harmless ignored results and meaningful values that are accidentally discarded.

Maybe also extend the check to `Promise.allSettled`.

**Comment by kirkwaiblinger** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/6631#issue-1622950997 it looks like
